File: packages/core/src/ratelimit/algorithms/slidingWindowCounter.ts
import type { Policy } from '../stores/Store.js';
import type { AlgorithmOutput } from './types.js';

export interface SlidingWindowCounterState {
  readonly windowStart: number;
  readonly previousCount: number;
  readonly currentCount: number;
}

/**
 * İki sabit pencerenin ağırlıklı ortalaması — doğruluk/maliyet dengesi,
 * yaklaşık sonuç verir (bkz. PLAN.md §4).
 */
export function slidingWindowCounter(
  state: SlidingWindowCounterState | undefined,
  policy: Policy,
  now: number,
  cost: number,
): AlgorithmOutput<SlidingWindowCounterState> {
  const windowStart = Math.floor(now / policy.windowMs) * policy.windowMs;
  const { previousCount, currentCount } = carryForward(state, windowStart, policy.windowMs);

  const elapsedInWindow = now - windowStart;
  const overlap = Math.max(0, (policy.windowMs - elapsedInWindow) / policy.windowMs);
  const estimate = previousCount * overlap + currentCount;

  const allowed = estimate + cost <= policy.limit;
  const ratePerMs = policy.limit / policy.windowMs;

  return {
    state: { windowStart, previousCount, currentCount: allowed ? currentCount + cost : currentCount },
    result: {
      allowed,
      limit: policy.limit,
      remaining: Math.max(0, Math.floor(policy.limit - (allowed ? estimate + cost : estimate))),
      retryAfterMs: allowed ? 0 : Math.ceil((estimate + cost - policy.limit) / ratePerMs),
    },
  };
}

function carryForward(
  state: SlidingWindowCounterState | undefined,
  windowStart: number,
  windowMs: number,
): { previousCount: number; currentCount: number } {
  if (!state) return { previousCount: 0, currentCount: 0 };
  if (state.windowStart === windowStart) return state;
  if (state.windowStart === windowStart - windowMs) {
    return { previousCount: state.currentCount, currentCount: 0 };
  }
  return { previousCount: 0, currentCount: 0 };
}

File: packages/core/src/ratelimit/lua/slidingWindowCounter.lua
-- KEYS[1] = key (hash: windowStart, currentCount, previousCount)
-- ARGV[1] = limit, ARGV[2] = windowMs, ARGV[3] = capacity (unused), ARGV[4] = now, ARGV[5] = cost
local limit = tonumber(ARGV[1])
local windowMs = tonumber(ARGV[2])
local now = tonumber(ARGV[4])
local cost = tonumber(ARGV[5])

local windowStart = math.floor(now / windowMs) * windowMs
local raw = redis.call('HMGET', KEYS[1], 'windowStart', 'currentCount', 'previousCount')
local storedWindowStart = tonumber(raw[1])
local currentCount = tonumber(raw[2]) or 0
local previousCount = tonumber(raw[3]) or 0

if storedWindowStart ~= windowStart then
  if storedWindowStart == windowStart - windowMs then
    previousCount = currentCount
  else
    previousCount = 0
  end
  currentCount = 0
end

local elapsedInWindow = now - windowStart
local overlap = math.max(0, (windowMs - elapsedInWindow) / windowMs)
local estimate = previousCount * overlap + currentCount

local allowed = (estimate + cost) <= limit
if allowed then
  currentCount = currentCount + cost
  estimate = previousCount * overlap + currentCount
end

redis.call('HSET', KEYS[1], 'windowStart', windowStart, 'currentCount', currentCount, 'previousCount', previousCount)
redis.call('PEXPIRE', KEYS[1], windowMs * 2)

local ratePerMs = limit / windowMs
local retryAfter = 0
if not allowed then
  retryAfter = math.ceil((estimate + cost - limit) / ratePerMs)
end

return { allowed and 1 or 0, math.max(0, math.floor(limit - estimate)), retryAfter }
